### PR TITLE
on Firefox choose screen sharing mode

### PR DESCRIPTION
**Pull request by hrotkogabor**
On Firefox, the user should choose what to share: full screen, or a single window.
Like here:
https://janus.conf.meetecho.com/screensharingtest.html

This pull request needs to be work together with the lib-jitsi-meet's firefox_share_fullscreen branch.

### CODE REVIEW

**Comment by kolAflash** [comment]
By the way, why is a Firefox add-on necessary to share the screen?
Firefox has an API for that (and Chrome probably too).

https://appear.in and https://spreed.me allow screen sharing without add-on and it works completely fine.

In case of https://spreed.me it's even possible to share single windows. There seems to be an API for that in Firefox, which you just need to call. Then Firefox will care about helping the user to select the windows he wants to share. (but sharing the whole screen should be possible too)

**Comment by bgrozev** [comment]
> By the way, why is a Firefox add-on necessary to share the screen?
> Firefox has an API for that

The API is only accessible from whitelisted domains. Appear.in and spreed.me (and meet.jit.si for that matter) are already whitelisted by mozilla, but other domain need to be added to the list before thay can access the API. This is what the extension does.

Similar situation with chrome, where the API is only accessible from an extension, which can be then queried by code executing in a tab.

> Then Firefox will care about helping the user to select the windows he wants to share.

The firefox UI for selecting a window is limited to selecting either from a list of windows, or a list of "the whole desktop", and not both. This is what this pull request addresses.

**Comment by hrotkogabor** [comment]
I have sent the signed CLA.

**Comment by kolAflash** [comment]
> The API is only accessible from whitelisted domains. Appear.in and spreed.me (and meet.jit.si for that matter) are already whitelisted by mozilla, but other domain need to be added to the list before thay can access the API. This is what the extension does.

Uhhh, ugly...

Found the Mozilla bug report about this.
https://bugzilla.mozilla.org/show_bug.cgi?id=1037424

What happens if the Jitsi Meet add-on modifies `media.getusermedia.screensharing.allowed_domains` ?
Especially if Mozilla adds other domains with the next Firefox update? Then this setting will be in the state "user modified" and will not be updated with further domains...

Very poor situation...
I'll take a closer look at the Mozilla bug report 1037424 and see why the hell they couldn't find a decent solution for that.

**Comment by hrotkogabor** [comment]
I have sent the signed CLA once again.

**Comment by damencho** [comment]
Jenkins: add to whitelist

File: conference.js [hunk 1]
```diff
@@ -1016,62 +1124,14 @@ export default {
...
-                        "dialog.error");
-                } else {
-                    dialogTxt = APP.translation.generateTranslationHTML(
-                        "dialog.failtoinstall");
-                    dialogTitle = APP.translation.generateTranslationHTML(
-                        "dialog.permissionDenied");
-                }
+            var ffShareMode = 'window';
+            if ( arguments.length == 0 && JitsiMeetJS.util.RTCBrowserType.isFirefox() ) {
+                askFirefoxScreensahringMode(this);
```
Comment (chrisdl) [inline conference.js]: Typo: Screensahring -> Screensharing

**Comment by hrotkogabor** [comment]
Could you help me, why do I have conflicts?

**Comment by maxdelorme** [comment]
Should be a very interesting feature

**Comment by tuxayo** [comment]
@hrotkogabor 
> Could you help me, why do I have conflicts?

Because you extracted code into doScreenShare() which is a good idea but meanwhile this code changed on upstream.
And also because some of the params of createLocalTracks () are now in an option object, so the addition of ffShareMode was conflicting.

It was a tough sync but I think I managed to solve the conflicts without breaking everything. Although I couldn't try to run the code due to lack of time. (I've never setup JitsiMeet)

In the function `createLocalTracks (options, checkForPermissionPrompt, ffShareMode)`
I didn't include ffShareMode in options because it was harder and I wasn't sure it's was the right thing to do.

So you have two pull request on your repo containing the sync with upstream/master
- https://github.com/hrotkogabor/jitsi-meet/pull/1
- https://github.com/hrotkogabor/jitsi-meet/pull/2

I tried with merge and rebase. The code is the same between the two branches (empty diff)
So use whichever you prefer to sync your branch with upstream/master.

About the rebase one, having your commit dated from june just after one from December is strange though. https://github.com/tuxayo/jitsi-meet/commits/firefox_share_fullscreen_solve_conflits_with_rebase

**Comment by tuxayo** [comment]
Now that the conflicts have been solved, what are the next steps toward getting this accepted into master?

Also, what went wrong with the tests? Is the PR content the cause?

**Comment by damencho** [comment]
The test reports rotate and I see the report for these tests had been deleted, but I see:
20:35:45 /home/jenkins/lib-jitsi-meet/JitsiMeetJS.js
20:35:45   389:9  error  Parsing error: Unexpected token AuthUtil
20:35:45 
20:35:45 /home/jenkins/lib-jitsi-meet/modules/RTC/ScreenObtainer.js
20:35:45   190:13  error  'self' is assigned a value but never used  no-unused-vars
20:35:45 
20:35:45 ✖ 2 problems (2 errors, 0 warnings)

I will re-run them now.

Jenkins test this please.

**Comment by damencho** [comment]
Yep, the same errors.

**Comment by hrotkogabor** [comment]
Thanx for the help, I started it over from upstream.
I also tested it in my environment, and fixed the latest 2 errors.
Please try again.
This PR is together with this:
https://github.com/jitsi/lib-jitsi-meet/pull/129

**Comment by bgrozev** [comment]
@hrotkogabor Some of the changes to conference.js don't pass jshint (and there might be more from eslint):

conference.js: line 233, col 90, Line is too long.
conference.js: line 319, col 90, Line is too long.
conference.js: line 345, col 90, Line is too long.
conference.js: line 1144, col 89, Line is too long.
conference.js: line 1144, col 35, Use '===' to compare with '0'.
conference.js: line 360, col 21, 'externalInstallation' is not defined.
conference.js: line 365, col 24, 'externalInstallation' is not defined.
conference.js: line 376, col 12, 'externalInstallation' is not defined.
conference.js: line 398, col 12, 'externalInstallation' is not defined.

You can run jshint+eslint locally with "npm run lint" (and there's a hook which is supposed to run them at every commit).

**Comment by hrotkogabor** [comment]
fixed the problems

**Comment by bgrozev** [comment]
Jenkins ok to test

File: conference.js [hunk 2]
```diff
@@ -298,6 +301,145 @@ function changeLocalDisplayName(nickname = '') {
     APP.UI.changeDisplayName(APP.conference.getMyUserId(), formattedNickname);
 }
 
+/**
+ * On firefox ask for destop share type: full screen or pick window.
+ * @param this named as that
+ * @returns {string} - Firefox screen share mode {window(default)|screen}
+ */
+function askFirefoxScreensharingMode (that) {
```
Comment (damencho) [inline conference.js]: The param 'that' is not used as doScreenShare do not have a second param.

File: conference.js [hunk 3]
```diff
@@ -298,6 +301,145 @@ function changeLocalDisplayName(nickname = '') {
...
+    );
+}
+
+/**
+ * do the screen share
+ * moved from this, to be able to handle confirmation
+ *
+ * @param {string|null} [ffShareMode] -
+ *       Firefox screen share mode {window(default)|screen}
+ * @param this
```
Comment (damencho) [inline conference.js]: No such parameter.

File: conference.js [hunk 4]
```diff
@@ -999,97 +1141,14 @@ export default {
...
-
-                APP.UI.messageHandler.openDialog(
-                    dialogTitleKey, dialogTxt, false);
-            });
+            var ffShareMode = 'window';
+            if ( arguments.length === 0 &&
+                        JitsiMeetJS.util.RTCBrowserType.isFirefox() ) {
+                askFirefoxScreensharingMode(this);
+            } else {
+                doScreenShare(ffShareMode, this);
```
Comment (damencho) [inline conference.js]: No second param.

File: conference.js [hunk 5]
```diff
@@ -999,97 +1141,14 @@ export default {
...
-                    dialogTitleKey = "dialog.permissionDenied";
-                }
-
-                APP.UI.messageHandler.openDialog(
-                    dialogTitleKey, dialogTxt, false);
-            });
+            var ffShareMode = 'window';
+            if ( arguments.length === 0 &&
+                        JitsiMeetJS.util.RTCBrowserType.isFirefox() ) {
+                askFirefoxScreensharingMode(this);
```
Comment (damencho) [inline conference.js]: No need of param.

File: lang/main.json [hunk 6]
```diff
@@ -311,6 +311,10 @@
         "micConstraintFailedError": "Yor microphone does not satisfy some of required constraints.",
         "micNotSendingData": "We are unable to access your microphone. Please select another device from the settings menu or try to restart the application.",
         "cameraNotSendingData": "We are unable to access your camera. Please check if another application is using this device, select another device from the settings menu or try to restart the application.",
+        "ffFullscreenShareQuestionTitle": "Share fullscreen, or a single window?",
```
Comment (damencho) [inline lang/main.json]: Should be 'full screen', to be consistent with the rest of the strings.

File: lang/main.json [hunk 7]
```diff
@@ -311,6 +311,10 @@
...
+        "ffFullscreenShareQuestion": "In firefox you should choose what to share.",
+        "ffFullscreenShareQuestionYes": "Share Full screen",
```
Comment (damencho) [inline lang/main.json]: Should be 'full screen', to be consistent with the rest of the strings.

**Comment by bgrozev** [comment]
Can we rename "ffShareMode" to something consistent with the rest of the config options for firefox screen sharing (e.g. desktopSharingFirefoxExtId)?

File: lang/main.json [hunk 8]
```diff
@@ -311,6 +311,10 @@
         "micConstraintFailedError": "Yor microphone does not satisfy some of required constraints.",
         "micNotSendingData": "We are unable to access your microphone. Please select another device from the settings menu or try to restart the application.",
         "cameraNotSendingData": "We are unable to access your camera. Please check if another application is using this device, select another device from the settings menu or try to restart the application.",
+        "ffFullscreenShareQuestionTitle": "Share fullscreen, or a single window?",
+        "ffFullscreenShareQuestion": "In firefox you should choose what to share.",
```
Comment (saghul) [inline lang/main.json]: Firefox (capitalization)

File: conference.js [hunk 9]
```diff
@@ -230,9 +230,11 @@ function maybeRedirectToWelcomePage(options) {
  *      undefined - one from settings will be used
  * @param {boolean} (checkForPermissionPrompt) - if lib-jitsi-meet should check
  *      for gUM permission prompt
+ * @param {string|null} [ffShareMode] -
+ *          Firefox screen share mode {window(default)|screen}
  * @returns {Promise<JitsiLocalTrack[]>}
  */
-function createLocalTracks (options, checkForPermissionPrompt) {
+function createLocalTracks (options, checkForPermissionPrompt, ffShareMode) {
```
Comment (saghul) [inline conference.js]: Can't this go into `options`? It looks kinda out of place IMHO.

File: conference.js [hunk 10]
```diff
@@ -298,6 +301,145 @@ function changeLocalDisplayName(nickname = '') {
     APP.UI.changeDisplayName(APP.conference.getMyUserId(), formattedNickname);
 }
 
+/**
+ * On firefox ask for destop share type: full screen or pick window.
+ * @param this named as that
+ * @returns {string} - Firefox screen share mode {window(default)|screen}
+ */
+function askFirefoxScreensharingMode (that) {
+    var title = "dialog.ffFullscreenShareQuestionTitle";
```
Comment (saghul) [inline conference.js]: Please, can you make all these `const` ?

File: conference.js [hunk 11]
```diff
@@ -298,6 +301,145 @@ function changeLocalDisplayName(nickname = '') {
...
+        msg,
+        true,
+        buttons,
+        function (e, submitValue) {
+
+            // Do not close the dialog yet
+            // e.preventDefault();
+
+            // Open dialog
+            var ffShareMode = 'window';
```
Comment (saghul) [inline conference.js]: let

**Pull request closed without merging**